File: l3_line_following/digital_twin/python/qlabs_setup.py
#-----------------------------------------------------------------------------#
#--------------------Quanser Interactive Labs Setup for-----------------------#
#---------------------------Mobile Robotics Lab-------------------------------#
#-----------------(Environment: QBot Platform / Warehouse)--------------------#
#-----------------------------------------------------------------------------#

import sys
from qvl.walls import QLabsWalls
from qvl.qlabs import QuanserInteractiveLabs
from qvl.qbot_platform import QLabsQBotPlatform
from qvl.qbot_platform_flooring import QLabsQBotPlatformFlooring
from qvl.real_time import QLabsRealTime
import pal.resources.rtmodels as rtmodels
import time
import numpy as np
import os
import subprocess

#------------------------------ Main program ----------------------------------

def setup(
        locationQBotP       = [0, 0, 0.1],
        rotationQBotP       = [0,0,90],
        verbose             = True,
        rtModel_workspace   = rtmodels.QBOT_PLATFORM,
        rtModel_driver      = rtmodels.QBOT_PLATFORM_DRIVER
        ):

    subprocess.Popen(['quanser_host_peripheral_client.exe', '-q'])
    time.sleep(2.0)
    subprocess.Popen(['quanser_host_peripheral_client.exe', '-uri', 'tcpip://localhost:18444'])
    
    qrt = QLabsRealTime()
    if verbose: print("Stopping any pre-existing RT models")
    qrt.terminate_real_time_model(rtModel_workspace)
    time.sleep(1.0)
    qrt.terminate_real_time_model(rtModel_driver)
    time.sleep(1.0)
    qrt.terminate_all_real_time_models()

    qlabs = QuanserInteractiveLabs()
    # Ensure that QLabs is running on your local machine
    if verbose: print("Connecting to QLabs...")
    if (not qlabs.open("localhost")):
        print("Unable to connect to QLabs")
        sys.exit()
        return
    if verbose: print("Connected to QLabs")

    qlabs.destroy_all_spawned_actors()

    #---------------------------- QBot Platform ---------------------------
    if verbose: print("Spawning QBot Platform ...")
    hQBot = QLabsQBotPlatform(qlabs)
    hQBot.spawn_id_degrees(actorNumber=0,
                        location=locationQBotP,
                        rotation=rotationQBotP,
                        scale=[1,1,1],
                        configuration=1,
                        waitForConfirmation= False)
    hQBot.possess(hQBot.VIEWPOINT_TRAILING)
    #------------------------------- Walls --------------------------------
    if verbose: print("Spawning walls ...")
    hWall = QLabsWalls(qlabs)
    hWall.spawn_degrees(location=[2, 1.2, 0.1], rotation=[0, 0, 0])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[2, 0, 0.1], rotation=[0, 0, 0])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[2, -1.2, 0.1], rotation=[0, 0, 0])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[-2, 1.2, 0.1], rotation=[0, 0, 0])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[-2, 0, 0.1], rotation=[0, 0, 0])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[-2, -1.2, 0.1], rotation=[0, 0, 0])

    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[1.2, 2, 0.1], rotation=[0, 0, 90])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[0, 2, 0.1], rotation=[0, 0, 90])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[-1.2, 2, 0.1], rotation=[0, 0, 90])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[1.2, -2, 0.1], rotation=[0, 0, 90])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[0, -2, 0.1], rotation=[0, 0, 90])
    hWall.set_enable_dynamics(True)
    hWall.spawn_degrees(location=[-1.2, -2, 0.1], rotation=[0, 0, 90])
    hWall.set_enable_dynamics(True)
    #----------------------------- Flooring -------------------------------
    if verbose: print("Spawning flooring ...")
    hFloor0 = QLabsQBotPlatformFlooring(qlabs)
    _ = hFloor0.spawn_id(actorNumber = 0,
                        location = [-0.6, 0.6,   0],
                        rotation = [0,0,-np.pi/2],
                        scale = [1,1,1],
                        configuration = 5,
                        waitForConfirmation= False)
    _ = hFloor0.spawn_id(actorNumber = 1,
                        location = [ 0.6, 1.8,   0],
                        rotation = [0,0,-np.pi/2],
                        scale = [1,1,1],
                        configuration = 0,
                        waitForConfirmation= False)
    _ = hFloor0.spawn_id(actorNumber = 2,
                        location = [ 1.8,-0.6,   0],
                        rotation = [0,0, np.pi],
                        scale = [1,1,1],
                        configuration = 0,
                        waitForConfirmation= False)
    _ = hFloor0.spawn_id(actorNumber = 3,
                        location = [-0.6,-1.8,   0],
                        rotation = [0,0, np.pi/2],
                        scale = [1,1,1],
                        configuration = 0,
                        waitForConfirmation= False)
    _ = hFloor0.spawn_id(actorNumber = 4,
                        location = [-1.8, 0.6,   0],
                        rotation = [0,0,0],
                        scale = [1,1,1],
                        configuration = 0,
                        waitForConfirmation= False)
    _ = hFloor0.spawn_id(actorNumber = 5,
                        location = [-0.6, 0.6,   0],
                        rotation = [0,0,0],
                        scale = [1,1,1],
                        configuration = 1,
                        waitForConfirmation= False)
    _ = hFloor0.spawn_id(actorNumber = 6,
                        location = [ 0.6, 0.6,   0],
                        rotation = [0,0,-np.pi/2],
                        scale = [1,1,1],
                        configuration = 1,
                        waitForConfirmation= False)
    _ = hFloor0.spawn_id(actorNumber = 7,
                        location = [ 0.6,-0.6,   0],
                        rotation = [0,0, np.pi],
                        scale = [1,1,1],
                        configuration = 1,
                        waitForConfirmation= False)
    _ = hFloor0.spawn_id(actorNumber = 8,
                        location = [-0.6,-0.6,   0],
                        rotation = [0,0, np.pi/2],
                        scale = [1,1,1],
                        configuration = 1,
                        waitForConfirmation= False)
    #----------------------------- RT Models -------------------------------
    if verbose: print("Starting RT models...")
    time.sleep(2)
    qrt.start_real_time_model(rtModel_workspace, userArguments=False)
    time.sleep(1)
    qrt.start_real_time_model(rtModel_driver, userArguments=True, additionalArguments="-uri tcpip://localhost:17098")
    if verbose: print('QLabs setup completed')
    return hQBot

if __name__ == '__main__':
    setup(locationQBotP       = [0, 0, 0.1], verbose=True)

File: l3_line_following/digital_twin/matlab/qlabs_setup.m
caseNum = 2; % 1, 2, 3, or 4

system('quanser_host_peripheral_client.exe -q');
pause(2)
system('quanser_host_peripheral_client.exe -uri tcpip://localhost:18444 &');

% MATLAB Path

newPathEntry = fullfile(getenv('QAL_DIR'), '0_libraries', 'matlab', 'qvl');
pathCell = regexp(path, pathsep, 'split');
if ispc  % Windows is not case-sensitive
  onPath = any(strcmpi(newPathEntry, pathCell));
else
  onPath = any(strcmp(newPathEntry, pathCell));
end

if onPath == 0
    path(path, newPathEntry)
    savepath
end

% Stop RT models
try
    qc_stop_model('tcpip://localhost:17000', 'qbot_platform_driver_virtual')
    pause(1)
    qc_stop_model('tcpip://localhost:17000', 'QBotPlatform_Workspace')
catch error
end
pause(1)

% QLab connection
qlabs = QuanserInteractiveLabs();
connection_established = qlabs.open('localhost');

if connection_established == false
    disp("Failed to open connection.")
    return
end
disp('Connected')
verbose = true;
num_destroyed = qlabs.destroy_all_spawned_actors();

% Flooring
hFloor0 = QLabsQBotPlatformFlooring(qlabs);
    % center
    hFloor0.spawn_id(0, [-0.6, 0.6,   0], [0,0,-pi/2], [1,1,1], 5, false); 
    % corners
    hFloor0.spawn_id(1, [ 0.6, 1.8,   0], [0,0,-pi/2], [1,1,1], 0, false);
    hFloor0.spawn_id(2, [ 1.8,-0.6,   0], [0,0, pi  ], [1,1,1], 0, false);
    hFloor0.spawn_id(3, [-0.6,-1.8,   0], [0,0, pi/2], [1,1,1], 0, false);
    hFloor0.spawn_id(4, [-1.8, 0.6,   0], [0,0,    0], [1,1,1], 0, false);
    % sides
    hFloor0.spawn_id(5, [-0.6, 0.6,   0], [0,0,    0], [1,1,1], 5, false);
    hFloor0.spawn_id(6, [ 0.6, 0.6,   0], [0,0,-pi/2], [1,1,1], 5, false);
    hFloor0.spawn_id(7, [ 0.6,-0.6,   0], [0,0, pi  ], [1,1,1], 5, false);
    hFloor0.spawn_id(8, [-0.6,-0.6,   0], [0,0, pi/2], [1,1,1], 5, false);

% Walls
hWall = QLabsWalls(qlabs, verbose);
    hWall.spawn_degrees([2, 1.2, 0.1], [0, 0, 0]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([2, 0, 0.1], [0, 0, 0]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([2, -1.2, 0.1], [0, 0, 0]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([-2, 1.2, 0.1], [0, 0, 0]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([-2, 0, 0.1], [0, 0, 0]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([-2, -1.2, 0.1], [0, 0, 0]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([1.2, 2, 0.1], [0, 0, 90]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([0, 2, 0.1], [0, 0, 90]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([-1.2, 2, 0.1], [0, 0, 90]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([1.2, -2, 0.1], [0, 0, 90]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([0, -2, 0.1], [0, 0, 90]);
    hWall.set_enable_dynamics(true);
    hWall.spawn_degrees([-1.2, -2, 0.1], [0, 0, 90]);
    hWall.set_enable_dynamics(true);

% QBot
hQBot = QLabsQBotPlatform(qlabs, verbose);
location = [0, 0, 0; -1.35, 0.3, 0; -1.5, 0, 0; -1.5, 0, 0];
rotation = [0, 0, 0;    0,   0, 0;   0, 0, 90;  0, 0, -90];
    hQBot.spawn_id_degrees(0, location(caseNum, :), rotation(caseNum, :), [1, 1, 1], 1) ;
    hQBot.possess(hQBot.VIEWPOINT_TRAILING);


    file_workspace = fullfile(getenv('RTMODELS_DIR'), 'QBotPlatform', 'QBotPlatform_Workspace.rt-win64');
    file_driver    = fullfile(getenv('RTMODELS_DIR'), 'QBotPlatform', 'qbot_platform_driver_virtual.rt-win64');


% Start RT models
pause(2)
system(['quarc_run -D -r -t tcpip://localhost:17000 ', file_workspace]);
pause(1)
system(['quarc_run -D -r -t tcpip://localhost:17000 ', file_driver, ' -uri tcpip://localhost:17098']);
pause(3)

